### ISSUE TITLE
"unable to provide IntelliSense configuration information" "peek definition"

### ISSUE THREAD

**Issue by enelson1001** [vector-of-bool/vscode-cmake-tools#586]
### Brief Issue Summary
Similar to issue #481. Everytime I open a header file from the project I get a popup-warning with "unable to provide IntelliSense configuration information".  I started new project and when the notification popup came up "CMake would like to configure intelliSense for this folder" I clicked allowed.
The #includes have squiggles
![image](https://user-images.githubusercontent.com/43481343/48881058-8844f080-edd0-11e8-8c7b-638fb457167b.png)
The build directory shows the smooth components being built.
![image](https://user-images.githubusercontent.com/43481343/48881210-597b4a00-edd1-11e8-8e8b-9adb92d88994.png)
The c_cpp_properties.json file that CMake Tools built did not include the compiler I had selected for the kit. 
![image](https://user-images.githubusercontent.com/43481343/48881307-f0480680-edd1-11e8-9b8b-7e0315d1c443.png)
I modified c_cpp_properties.json to include the correct compiler and named it Linux-ESP32
![image](https://user-images.githubusercontent.com/43481343/48881348-22f1ff00-edd2-11e8-9cd2-47eb2e1d3bf4.png)

### Expected:
Thought from this explanation https://vector-of-bool.github.io/2018/07/20/cmt-1.1.0.html everything would be simplified and I wouldn't have to add the all the includes paths for the esp-idf components.

At one point I did not have the squiggly red lines but if I hoovered over the include statement and right clicked and select for example "peek definition" VSCode would respond file not found.

I should add that I do not receive any compile errors and that I am able to flash the ESP32 device and the program runs properly.

So the questions is
1. Should I not use use "CMake would like to configure intelliSense for this folder" and write my own c_cpp_properties.json will every include in the esp-idf/components folder?
2. If there is some mistake I made configuring the project and the use of "CMake would like to configure intelliSense for this folder" actually works, does hoovering over the include allow me to bring up the definition?

### Platform and Versions

- **Operating System**:  Linux Mint
- **CMake Version**:  3.12
- **VSCode Version**:  1.29.1
- **CMake Tools Extension Version**:  1.1.2
- **Compiler/Toolchain**:  xtensa-esp32-elf

**Comment by Randshot** [comment]
Could you show us your CMakeLists.txt
CMake has to know the include paths, so the C++ can use that information for intelli sense.

**Comment by enelson1001** [comment]
*Here is the top level CMakeLists.txt file*
# The following lines of boilerplate have to be in your project's
# CMakeLists in this exact order for cmake to work correctly

# This line must be the first line in the CMakeLists.txt file.
# ESP-IDF is designed to work with CMake 3.5 or newer.
cmake_minimum_required(VERSION 3.10)

# Use -std=c++14
set(CMAKE_CXX_STANDARD 14)

# For std::to_string (until using xtensa-gcc 8)
add_definitions(-D_GLIBCXX_USE_C99)

# Pulls in the rest of the CMake functionality to configure
# the project, discover all the components, etc.
include($ENV{IDF_PATH}/tools/cmake/project.cmake)

# The project name is used for the final binary output files
# of the app - ie myProject.elf, myProject.bin
project(GreetingCPPv2)



*Here is the CMakeLists.txt file inside main directory*
# ESP-IDF template for a component. main is now treated as a component


# COMPONENT_REQUIRES is a (space-separated) list of components that
# are required to include this project's header files into other
# components. If this component has a header file in a
# COMPONENT_ADD_INCLUDEDIRS directory that includes a header from
# another component, that component should be listed in
# COMPONENT_REQUIRES. Requirements are recursive.

# set(COMPONENT_REQUIRES smooth json)

# If a component only requires another component’s headers to
# compile its source files (not for including this component’s headers),
# then these components should be listed in COMPONENT_PRIV_REQUIRES instead.

# set(COMPONENT_PRIV_REQUIRES)

# COMPONENT_SRCS is a (space-separated) list of source files
# (*.c, *.cpp, *.cc, *.S). These source files will be compiled
# into the component library.
set(COMPONENT_SRCS ATask.cpp BTask.cpp GreetingCPPv2.cpp  main.cpp)


# COMPONENT_ADD_INCLUDEDIRS is a (space-separated) list of directories
# to add to the global include search path for any component which
# requires this component, and also the main source files.
# PROJECT_PATH: Absolute path of the project directory containing
# the project CMakeList.txt. Same as the CMAKE_SOURCE_DIR variable.

set(COMPONENT_ADD_INCLUDEDIRS "${CMAKE_SOURCE_DIR}/include")

# register_component() is required to add the component (using the
# variables set above) to the build. A library with the name of the
# component will be built and linked into the final app. If this step
# is skipped, this component will not be part of the build.

register_component()




Thanks

**Comment by vector-of-bool** [comment]
This is a common thing that trips people up. The problem is that the header files are not being added to the target, so CMake Tools has no knowledge of what target is being used with the header files. Try adding the header files to your `COMPONENT_SRCS` variable (CMake _should_ be smart enough to not try and "compile" header files).

Also, if you let CMake Tools provide the configuration information, the correct compiler will be given to cpptools, and `c_cpp_properties.json` won't be necessary for files that CMake Tools knows about.

**Comment by vector-of-bool** [comment]
The header files in which you are seeing the red-underlines are the header files that need to be added to the source list. In your example, the `HelloCPP.h` file needs to be added to the source list of a target.

**Comment by enelson1001** [comment]
OK this is the steps I took
1. Renamed c_cpp_properties.json to fred_c_cpp_properties.json. So file the c_cpp_properties.json file does not exist.
2.There are two CMakeLists.txt files .  I did not touch the top level CMakeLists.txt file. The CMakeLists.txt file I modified was under the "main" directory.  I changed the following in the file from set(COMPONENT_SRCS HelloCPP.cpp main.cpp) to set(COMPONENT_SRCS HelloCPP.cpp ../include/HelloCPP.h main.cpp).  So I added the HelloCPP.h file.
3. Selected CMAKE icon (left side menu).  Clicked on the 3 dots(upper right side. Selected "clean reconfigure".
4. Went back to the menu at the bottom of the screen and clicked "Build [all].
5. The build competed with no errors.
6. HelloCPP.h has the same red squiggly lines as before - no change.

I assumed step 3 would give me a clean slate for project.

Thanks

**Comment by vector-of-bool** [comment]
Interesting... I took a look [at `register_component` defined here](https://github.com/espressif/esp-idf/blob/master/tools/cmake/components.cmake#L17), and it looks like it should work...

A few more diagnosis steps:

1 In the project outline panel, under the target for your component, do you see `HelloCPP.h` listed as a source file for the target?

2. Where do you expect `smooth/core/Application.h` to resolve to (what is the file's full path)?

3. Open `HelloCPP.cpp` and run the `CMake: Compile File` from the command from the palette (Ctrl+Shift+P). This should open a terminal and execute the command that will compile that particular file. Paste that command here.

4. Do you still see the "unable to provide IntelliSense configuration information" after making the modifications? (May need to reload the window)

**Comment by enelson1001** [comment]
I was able to HelloCPP project to build without errors and also show no red
squiggly lines and without needing c_cpp_properties.json file.  I created a
Github repository that has the HelloCPP project and a new Project
GreetingCPP.  The github repo is at
https://github.com/enelson1001/GreetingCPP

Under vscode-workspace-esp32 directory is a Getting-Started.txt that lists
the steps I took to create a project
Under externals/smooth is a Readme.txt file that explains what I did to get
the HelloCPP to compile.

But then I tried another project GreetingCPP.  I am getting build errors:
[build] FAILED: main/CMakeFiles/main.dir/GreetingCPPv2.cpp.obj
[build] In file included from ../main/GreetingCPPv2.cpp:5:0:
[build] ../include/GreetingCPPv2.h:3:38: fatal error:
smooth/core/Appliccation.h: No such file or directory
[build] compilation terminated.

Even though the externals/smooth directory has everything.  I did not
upload these files to repository.  They can be found here.
https://github.com/PerMalmberg/Smooth.git

This is probably not a CMake Tools problem and you can close the issue if
you like but I thought I would ask - Do you have any idea why it can't find
Application.h?

Thanks

**Comment by vector-of-bool** [comment]
It looks to be just a misspelling of `Application` as `Appliccation`

**Comment by enelson1001** [comment]
Thanks that was a dumb mistake.

I am not back to HelloCPP project.  If I leave the
externals/smooth/lib/smooth/CMakeLists.txt as original then CMake throws
errors.
I corresponded with the author of Smooth library and he said he can build
my application without any problems but he is using Clion.
Under github https://github.com/enelson1001/GreetingCPP I added a folder
called CMake-Problems that has some files.
1. The-Ouput-Console-Log-Messages-With-Errors.txt  --- shows the errors I
am getting along with all the additional message I added to
lib/smooth/CMakeList.txt and project.cmake
2. The-modified-project-cmake-file-with-messages.txt  --- shows additional
messages added for debugging
3.
The-modified-externals-smooth-lib-smooth-CMakeLists-file-with-messages.txt
--- shows the additional messages I added for debugging

The-Ouput-Console-Log-Messages-With-Errors.txt
1. Line 132 - When I look at the build directory there is no file called
/home/ed/vscode-workspace-esp32/sandbox/HelloCPP/build/esp32/esp32_out.ld??
2. Lines 142 - 145 - This should not happen this is an ESP_PLATFORM as seen
in Line 11. How can this happen???
3. Lines 159-172 CMake errors is because of maybe lines 142-145?

Thanks
